sdk/core/azure-core/test/ut/operation.cpp

```diff
@@ -0,0 +1,56 @@
+// ************ CUSTOMER CODE **********************
+
+#include "gtest/gtest.h"
+
+#include "operation_test.hpp"
+
+#include <azure/core/context.hpp>
+#include <azure/core/operation.hpp>
+
+#include <chrono>
+
+using namespace Azure::Core::Test;
+using namespace std::literals;
+
+TEST(Operation, Poll)
+{
+  StringClient client;
+  auto operation = client.StartStringUpdate();
+
+  EXPECT_FALSE(operation.Done());
+  EXPECT_FALSE(operation.HasValue());
+
+  while(!operation.Done())
+  {
+    EXPECT_FALSE(operation.HasValue());
+    auto response = operation.Poll();
+  }
+
+  EXPECT_TRUE(operation.Done());
+  EXPECT_TRUE(operation.HasValue());
+
+  auto result = operation.Value();
+  EXPECT_TRUE(result == "StringOperation-Completed");
+}
+
+TEST(Operation, PollUntilDone)
+{
+  StringClient client;
+  auto operation = client.StartStringUpdate();
+
+  EXPECT_FALSE(operation.Done());
+  EXPECT_FALSE(operation.HasValue());
+  
+  auto start = std::chrono::high_resolution_clock::now();
+  auto response = operation.PollUntilDone(500ms);
+  auto end = std::chrono::high_resolution_clock::now();
+  std::chrono::duration<double, std::milli> elapsed = end - start;
+  //StringOperation test code is implemented to poll 2 times
+  EXPECT_TRUE(elapsed >= 1s);
+
+  EXPECT_TRUE(operation.Done());
+  EXPECT_TRUE(operation.HasValue());
```

Comment:
What happens if we call `PollUntilDone` again after operation is done? I would imagine that should return almost right away (or one first poll) and keep the state unchanged. Consider adding that test.

Reply:
Its an implementation detail for the operation

Reply:
It should return immediately.
If we leave too much to be an "implementation detail" then we will get different implementations and this would be good for customers. I think we should err on the side of being prescriptive here with behavior and put it in our guidelines.

Also, the more code we put in the base class and the less code client library devs have to write in the derived classes helps ensure this while making it easier for client lib devs to implement thier derived class.

Reply:
> then we will get different implementations and this would be good for customers. 

I am assuming, @JeffreyRichter, that you meant: this would **not** be good for customers.